### PR TITLE
netcdf-c: An incorrect hdf5 library name is added to pkconfig and cmake config 

### PR DESCRIPTION
**Pull request by Chrismarsh**
When netcdf-c is built with cmake, hdf5-shared and hdf5_hl-shared are added as library names to the pkconfig and cmake config files. This impacts consuming libraries, such as vtk, that go onto have nonexistent library names in the link path.

This probably impacts more than just netcdf-c 4.9.2; however, I don't have the bandwidth to test other versions

I suspect this PR also resolves:
https://github.com/spack/spack/issues/42695

Referencing upstream issues:
https://github.com/Unidata/netcdf-c/issues/2869
https://github.com/Unidata/netcdf-c/issues/1444#issuecomment-1967270176 

Likely fixed in on-going netcdf-c upstream work:
https://github.com/Unidata/netcdf-c/pull/2847

### CODE REVIEW

**Review by scheibelp — DISMISSED** [review]
> I suspect this PR also resolves: https://github.com/spack/spack/issues/42695

This may resolve the initially reported issue, although that issue also describes RPATH problems (and also was later unable to reproduce the library location problem). In any case, this looks good whether it resolves that issue.

Also from reading through some threads you link, it seems that `hdf5` library naming conventions might change if `ONLY_SHARED_LIBS=1` (we always turn it off in the builtin repo for `hdf5`).

I'm willing to merge as-is, but I will wait a day to see if the maintainers have opinions on this.

**Comment by Chrismarsh** [comment]
>Also from reading through some threads you link, it seems that hdf5 library naming conventions might change if ONLY_SHARED_LIBS=1

I saw that too. I made the change to hdf5 and rebuilt `hdf5`. However, that change doesn't seem to change anything for me:

```
$ ag hdf5-shared
cmake/hdf5-config.cmake
90:set (${HDF5_PACKAGE_NAME}_EXPORT_LIBRARIES     hdf5-shared;hdf5_tools-shared)

cmake/hdf5-targets.cmake
22:foreach(_cmake_expected_target IN ITEMS hdf5-shared mirror_server mirror_server_stop hdf5_tools-shared h5diff-shared ph5diff-shared h5ls-shared h5debug-shared h5repart-shared h5mkgrp-shared h5clear-shared h5delete-shared h5import-shared h5repack-shared h5jam-shared h5unjam-shared h5copy-shared h5stat-shared h5dump-shared h5format_convert-shared h5perf)
56:# Create imported target hdf5-shared
57:add_library(hdf5-shared SHARED IMPORTED)
59:set_target_properties(hdf5-shared PROPERTIES
77:  INTERFACE_LINK_LIBRARIES "hdf5-shared"

cmake/hdf5-targets-release.cmake
8:# Import target "hdf5-shared" for configuration "Release"
9:set_property(TARGET hdf5-shared APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
10:set_target_properties(hdf5-shared PROPERTIES
15:list(APPEND _cmake_import_check_targets hdf5-shared )
16:list(APPEND _cmake_import_check_files_for_hdf5-shared "${_IMPORT_PREFIX}/lib/libhdf5.so.310.3.0" )
```

>(and also was later unable to reproduce the library location problem)

I've had a few of those. In this case, it only happens if:
- hdf5 is built with cmake
- netdf-c is built with cmake
- netcdf-c uses the hdf5 cmake config file

Given netcdf-c supports both autotools and cmake, if `netcdf-c build_system=cmake` is not selected, it won't show up.  So I can imagine something changing and suddenly cannot be reproduced.

File: var/spack/repos/builtin/packages/netcdf-c/package.py [hunk 1]
```diff
@@ -65,6 +65,9 @@ class NetcdfC(CMakePackage, AutotoolsPackage):
         # the changes not incorporated into that PR
         patch("netcdfc_correct_and_export_link_interface.patch", when="platform=windows")
 
+        # Incorrect hdf5 libraries names are put in the package config and config.cmake files
+        patch("4.9.2-fix-hdf5-pkgconfigcmake.patch", when="@4.9.2")
```
Comment (skosukhin) [inline var/spack/repos/builtin/packages/netcdf-c/package.py]:
```suggestion
        patch("4.9.2-fix-hdf5-pkgconfigcmake.patch", when="@4.9.2 build_system=cmake")
```

**Review by skosukhin — CHANGES_REQUESTED** [review]
1. I think it makes sense to replace the patch file with a `@run_after("install")` method that fixes the required files with the `filter_file` method.
2. I would remove the `diff --git ...` and `index ...` lines from the patch files.
3. I don't think that replacing targets with linker flags in the CMake files is a good idea. Instead, the config file should call `find_dependency` to make the targets available to the consuming packages.

**Comment by Chrismarsh** [comment]
To be clear: this is an extension of how netcdf already handles fixing up the target name.

Specifically Re:#3 I agree, but all that target patching already exists in the netcdf file. I only extended it to also include the extra name. The upstream work I linked looks like they're fixing this properly.

Re#1, this was an extension of the existing pkgconfig patching applied to the cmake config.

I chose this way to keep it in line with how the package already handled it. However, I can change this to patching the output files later today.

**Comment by Chrismarsh** [comment]
@skosukhin I've migrated to what you suggest via a post-install filter_file approach.

File: var/spack/repos/builtin/packages/netcdf-c/package.py [hunk 2]
```diff
@@ -351,6 +351,22 @@ def cmake_args(self):
...
+        https://github.com/spack/spack/pull/42878
+        """
+
+        pkgconfig_file = join_path(self.prefix.lib64.pkgconfig, "netcdf.pc")
+        filter_file("hdf5-shared", "hdf5", pkgconfig_file)
+        filter_file("hdf5_hl-shared", "hdf5_hl", pkgconfig_file)
+
+        cmakeconfig_file = join_path(self.prefix.lib64.cmake.netCDF, "netCDFTargets.cmake")
+        filter_file("hdf5-shared", "hdf5", cmakeconfig_file)
+        filter_file("hdf5_hl-shared", "hdf5_hl", cmakeconfig_file)
```
Comment (skosukhin) [inline var/spack/repos/builtin/packages/netcdf-c/package.py]: Thank you. Are you sure `bin/nc-config` doesn't need patching? Are you sure that the `*.pc` and `*Targets.cmake` files get installed to `lib64` on all platforms and never to `lib`? I would use [`find`](https://github.com/spack/spack/blob/983422facf5c4c7af9ef3456ff7d7f43ffa13305/lib/spack/llnl/util/filesystem.py#L1789) and [`FileFilter`](https://github.com/spack/spack/blob/983422facf5c4c7af9ef3456ff7d7f43ffa13305/lib/spack/llnl/util/filesystem.py#L439) here.

Comment (Chrismarsh) [reply var/spack/repos/builtin/packages/netcdf-c/package.py]: > `nc-config`

As built for me, it doesn't report the hdf5 libraries. Indeed, I think this is a bug, but that's a separate issue
```
 ./nc-config --libs
-L/home/chm003/project-ords/spack/opt/spack/linux-rhel8-icelake/gcc-9.3.0/netcdf-c-4.9.2-6fpvnbovjwauzxdtjfdeliopq6ybu4zz/lib64 -lnetcdf
```
however, you're right that `-shared` was still lurking inside the script. `.settings` was also impacted. I've also removed the lib64 assumption.

**Review by skosukhin — APPROVED** [review]

**Review by scheibelp — APPROVED** [review]

**Pull request merged**